File: db4on/core/src/core/com/db4o/MReadBytes.cs
namespace com.db4o
{
	internal sealed class MReadBytes : com.db4o.MsgD
	{
		internal sealed override com.db4o.YapWriter GetByteLoad()
		{
			int address = this._payLoad.ReadInt();
			int length = this._payLoad.GetLength() - (com.db4o.YapConst.INT_LENGTH);
			this._payLoad.RemoveFirstBytes(com.db4o.YapConst.INT_LENGTH);
			this._payLoad.UseSlot(address, length);
			return this._payLoad;
		}

		internal sealed override com.db4o.MsgD GetWriter(com.db4o.YapWriter bytes)
		{
			com.db4o.MsgD message = this.GetWriterForLength(bytes.GetTransaction(), bytes.GetLength
				() + com.db4o.YapConst.INT_LENGTH);
			message._payLoad.WriteInt(bytes.GetAddress());
			message._payLoad.Append(bytes._buffer);
			return message;
		}

		internal sealed override bool ProcessMessageAtServer(com.db4o.foundation.network.YapSocket
			 sock)
		{
			com.db4o.YapStream stream = GetStream();
			int address = this.ReadInt();
			int length = this.ReadInt();
			lock (stream.i_lock)
			{
				com.db4o.YapWriter bytes = new com.db4o.YapWriter(this.GetTransaction(), address, 
					length);
				try
				{
					stream.ReadBytes(bytes._buffer, address, length);
					GetWriter(bytes).Write(stream, sock);
				}
				catch (System.Exception e)
				{
					com.db4o.Msg.NULL.Write(stream, sock);
				}
			}
			return true;
		}
	}
}

File: db4oj/core/src/com/db4o/cs/messages/MReadBytes.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.cs.messages;

import com.db4o.*;
import com.db4o.cs.*;

public final class MReadBytes extends MsgD {
	public final YapReader getByteLoad() {
		int address = this._payLoad.readInt();
		int length = this._payLoad.getLength() - (YapConst.INT_LENGTH);
		this._payLoad.removeFirstBytes(YapConst.INT_LENGTH);
		this._payLoad.useSlot(address, length);
		return this._payLoad;
	}

	public final MsgD getWriter(YapWriter bytes) {
		MsgD message =
			this.getWriterForLength(bytes.getTransaction(), bytes.getLength() + YapConst.INT_LENGTH);
		message._payLoad.writeInt(bytes.getAddress());
		message._payLoad.append(bytes._buffer);
		return message;
	}
	
	public final boolean processAtServer(YapServerThread serverThread) {
	    YapStream stream = getStream();
		int address = this.readInt();
		int length = this.readInt();
		synchronized (stream.i_lock) {
			YapWriter bytes =
				new YapWriter(this.getTransaction(), address, length);
			try {
				stream.readBytes(bytes._buffer, address, length);
				serverThread.write(getWriter(bytes));
			} catch (Exception e) {
				// TODO: not nicely handled on the client side yet
				serverThread.write(Msg.NULL);
			}
		}
		return true;
	}
}